lightning/src/ln/channelmanager.rs

```diff
@@ -4498,12 +4589,91 @@ impl_writeable_tlv_based!(HTLCPreviousHopData, {
 	(6, incoming_packet_shared_secret, required)
 });
 
-impl_writeable_tlv_based!(ClaimableHTLC, {
-	(0, prev_hop, required),
-	(2, value, required),
-	(4, payment_data, required),
-	(6, cltv_expiry, required),
-});
+impl Writeable for ClaimableHTLC {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		let payment_data = match &self.claim_type {
+			ClaimableType::Invoice(data) => data.clone(),
+			_ => msgs::FinalOnionHopData {
+				payment_secret: PaymentSecret([0; 32]),
+				total_msat: 0,
+			}
+		};
+		let keysend_preimage = match self.claim_type {
+			ClaimableType::Invoice(_) => None,
+			ClaimableType::Spontaneous(preimage) => Some(preimage.clone()),
+		};
+		write_tlv_fields!
+		(writer,
+		 {
+		   (0, self.prev_hop, required), (2, self.value, required),
+		   (4, payment_data, required), (6, self.cltv_expiry, required),
+			 (8, keysend_preimage, option),
+		 });
+		Ok(())
+	}
+
+	#[inline]
+	fn serialized_length(&self) -> usize {
+		use util::ser::BigSize;
+		let len = {
+			#[allow(unused_mut)]
+			let mut len = ::util::ser::LengthCalculatingWriter(0);
+			get_varint_length_prefixed_tlv_length!(len, 0, self.prev_hop, required);
+			get_varint_length_prefixed_tlv_length!(len, 2, self.value, required);
+			let payment_data = match &self.claim_type {
+				ClaimableType::Invoice(data) => data.clone(),
+				_ => msgs::FinalOnionHopData {
+					payment_secret: PaymentSecret([0; 32]),
+					total_msat: 0,
+				}
+			};
+			get_varint_length_prefixed_tlv_length!(len, 4, payment_data, required);
+			get_varint_length_prefixed_tlv_length!(len, 6, self.cltv_expiry, required);
+			let keysend_preimage = match self.claim_type {
+				ClaimableType::Invoice(_) => None,
+				ClaimableType::Spontaneous(preimage) => Some(preimage.clone()),
+			};
+			get_varint_length_prefixed_tlv_length!(len, 8, keysend_preimage, option);
+			len.0
+		};
+		let mut len_calc = ::util::ser::LengthCalculatingWriter(0);
+		BigSize(len as u64).write(&mut len_calc).expect("No in-memory data may fail to serialize");
+		len + len_calc.0
+	}
+}
+
+impl Readable for ClaimableHTLC {
+	fn read<R: Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		init_tlv_field_var!(prev_hop, required);
+		init_tlv_field_var!(value, required);
+		init_tlv_field_var!(payment_data, required);
+		init_tlv_field_var!(cltv_expiry, required);
+		init_tlv_field_var!(keysend_preimage, option);
```

Comment:
nit: I kinda preferred these to be "internal" macros as utilities. If we're writing it out anyway we might as well not use the wrapper these macros create and just the objects explicitly instead. That would make this code (and the code below) much more readable, IMO.

Reply:
Ok, I didn't see an obvious way around using `OptionDeserWrapper`s. Lmk if the new version is off

Reply:
For types that it makes sense for (eg primitive types), you can just set a dummy value. eg just set integers to zero.